include/revng/Yield/Support/GraphLayout/SugiyamaStyle/InternalGraph.h

```diff
@@ -0,0 +1,234 @@
+  }
+
+  layout::Path &getPath() {
+    revng_assert(!isVirtual(), "`getPath` should only be used on real edges.");
+    return getPathImpl();
+  }
+
+private:
+  static void appendPointImpl(layout::Path &Path, const layout::Point &Point) {
+    if (Path.size() > 1) {
+      auto &First = *std::prev(std::prev(Path.end()));
+      auto &Second = *std::prev(Path.end());
+
+      auto LHS = (Point.Y - Second.Y) * (Second.X - First.X);
+      auto RHS = (Second.Y - First.Y) * (Point.X - Second.X);
+      if (LHS == RHS)
+        Path.pop_back();
+    }
+    Path.push_back(Point);
+  }
+
+public:
+  /// Adds a point to the edge path or modifies its last point if that's
+  /// sufficient, based on their coordinates.
+  void appendPoint(const layout::Point &Point) {
+    appendPointImpl(getPathImpl(), Point);
+  }
+  void appendPoint(layout::Coordinate X, layout::Coordinate Y) {
+    return appendPoint(layout::Point(X, Y));
+  }
+};
+
+} // namespace detail
+
+using InternalNode = MutableEdgeNode<detail::InternalNode,
+                                     detail::InternalEdge,
+                                     false>;
+
+class InternalGraph : public GenericGraph<InternalNode, 16, true> {
+public:
+  using GenericGraph<InternalNode, 16, true>::GenericGraph;
+
+private:
+  std::size_t NodeIndexCounter = 0;
+  std::size_t EdgeIndexCounter = 0;
+
+public:
+  template<typename NodeRef, typename EdgeRef>
+  struct OutputLookups {
+    std::vector<NodeRef> Nodes;
+    std::vector<EdgeRef> Edges;
+  };
+
+public:
+  template<layout::HasLayoutableGraphTraits GraphType>
+  static auto make(const GraphType &Graph) {
+    using LLVMTrait = llvm::GraphTraits<GraphType>;
+    using LayoutTrait = layout::LayoutableGraphTraits<GraphType>;
+
+    using NodeRef = typename LLVMTrait::NodeRef;
+    using EdgeRef = typename LLVMTrait::EdgeRef;
+
+    InternalGraph Result;
+    OutputLookups<NodeRef, EdgeRef> Lookup;
+    Lookup.Nodes.reserve(LLVMTrait::size(Graph));
+
+    std::unordered_map<NodeRef, InternalGraph::Node *> InternalLookup;
+    for (NodeRef Node : llvm::nodes(Graph)) {
+      layout::Size Size = LayoutTrait::getNodeSize(Node);
+
+      auto [It, Success] = InternalLookup.try_emplace(Node,
+                                                      Result.makeNode(Size));
+      revng_assert(Success);
+      revng_assert(It->second->Index == Lookup.Nodes.size());
+      Lookup.Nodes.emplace_back(Node);
+    }
+
+    for (NodeRef From : llvm::nodes(Graph)) {
+      for (EdgeRef Edge : llvm::children_edges<GraphType>(From)) {
+        const auto &NewEdge = Result.makeEdge(InternalLookup.at(From),
+                                              InternalLookup.at(Edge.Neighbor));
+        revng_assert(NewEdge->Index == Lookup.Edges.size());
+        Lookup.Edges.emplace_back(Edge);
+      }
+    }
+
+    struct ResultType {
+      InternalGraph Graph;
+      const OutputLookups<NodeRef, EdgeRef> Lookup;
+    };
```

Comment:
Give more dignity to this `struct` definition. It's part of your public API, you can't just hide the definition inside here, in a function whose return type is `auto`. Drop `auto` and make this struct visible outside. Perhaps `InternalGraph` itself should have a `Lookup` field and `ResultType` (or whatever you want to call it) should just go away.

Reply:
Ah, wait, I'm now seeing why I went with such an interior approach: since this struct includes the `InternalGraph` object as a member, it cannot be declared before the `InternalGraph` is.

Either way, please take a look at it now (`InternalGraphWithLookupTable`).

Reply:
> Perhaps `InternalGraph` itself should have a `Lookup` field

The entire point of this PR is to move the information `Lookup` represents outside of the `InternalGraph`. This way `InternalGraph` has no relation to whatever form the data is represented in (as in, it doesn't need to be a template - because `Lookup` is), and the graph itself can just refer to specific nodes by their indices in the `Lookup`.

Reply:
See my other new comment of `ResultType`. If you agree with that and do it, you can then close both conversations.

Reply:
Sadly that wouldn't work, see the response to the other comment.

P. S. Since this "- see the other comment", "- no, YOU see the other comment" approach kind of sucks, I'll close this thread and let the other one track this issue.